### ISSUE TITLE
error in running evaluation function

### ISSUE THREAD

**Issue by mohamedelbeih** [rojagtap/abstractive_summarizer#9]
i'm getting this error when trying to use summarize function , can you please help ??
thanks in advance
![image](https://user-images.githubusercontent.com/74533719/115181935-4a4f7200-a0d9-11eb-916c-6a84948620f6.png)

**Comment by luciusgustavo** [comment]
Hi, @mohamedelbeih! 
I was facing the same problem, so i rewrote the code
`document = document.apply(lambda x: "<go> " + x + "<stop>")
document.head()`
manually (insted of copy and past) and executed its dependent cells.  This solved the problem for me.

**Comment by rojagtap** [comment]
> Hi, @mohamedelbeih!
> I was facing the same problem, so i rewrote the code
> `document = document.apply(lambda x: "<go> " + x + "<stop>")`
> `document.head()`
> manually (insted of copy and past) and executed its dependent cells.
> This solved the problem for me.

Yes, I suppose this could be a formatting issue.

**Comment by SumanthSrungavarapu** [comment]
in the code this is done for summary ( summary = summary.apply(lambda x: "<go> " + x + "<stop>")).
So we should do this for both document and summary or only document or only summary
everything is running error free, but not getting expected results atleast near to you. please help
My notebook: https://colab.research.google.com/drive/1JR7bZb5gsmC9zphD-z8oX1gwJVmPk2jm?usp=share_link

**Comment by rojagtap** [comment]
> in the code this is done for summary ( summary = summary.apply(lambda x: " " + x + "")). So we should do this for both document and summary or only document or only summary everything is running error free, but not getting expected results atleast near to you. please help My notebook: https://colab.research.google.com/drive/1JR7bZb5gsmC9zphD-z8oX1gwJVmPk2jm?usp=share_link

Hi,

You just need to apply this to the decoder inputs (summary). The idea behind this is, the decoder inputs are something that we wouldn't know during inference. For instance, while training we know the expected summary, hence we simply use it with an offset `1`. But during inference, we need to give the model some indication that it needs to use the original document and generate a word for the summary. So it uses your document and the word `"<go>"` and generates the first word (say `"the"`). Then it takes the document and the context for the words `"<go>"` and `"the"` and generates the next word and so on. We know where to stop thanks to the `"<stop>" `keyword.

As far as the results are concerned, if you notice I have trained the model for 20 epochs, not 10 which is not clear in the notebook. I had to train it twice due to the runtime restrictions of colab. Maybe that is the reason for your issue.

**Comment by SumanthSrungavarapu** [comment]
Thanks for answering, now i started running it applying only for decoder inputs, its still running but my loss started with 10.5 very farther from yours. Few differences I observed are following
In Tokenizing the texts into integer tokens, your encoder_vocab_size, decoder_vocab_size are (76362, 29661), mine are (76362, 37307)
There are other small diffreneces in summary_lenghts properties (.describe()).
Am i doing something wrong or is all ok, thank you and please help.
collab link: https://colab.research.google.com/drive/1JR7bZb5gsmC9zphD-z8oX1gwJVmPk2jm?usp=share_link

**Comment by rojagtap** [comment]
> Thanks for answering, now i started running it applying only for decoder inputs, its still running but my loss started with 10.5 very farther from yours. Few differences I observed are following In Tokenizing the texts into integer tokens, your encoder_vocab_size, decoder_vocab_size are (76362, 29661), mine are (76362, 37307) There are other small diffreneces in summary_lenghts properties (.describe()). Am i doing something wrong or is all ok, thank you and please help. collab link: https://colab.research.google.com/drive/1JR7bZb5gsmC9zphD-z8oX1gwJVmPk2jm?usp=share_link

As I said earlier, I had trained the model earlier on top of which I trained it for the results displayed. I mean,
```
Epoch 1 Loss 2.5077
```

is after 10-20 epochs. I trained in the background, but it cannot be seen there.

**Comment by SumanthSrungavarapu** [comment]
ok I will continue the training.
so the change in decoder vocab size is not a problem? (orginal: 29661,mine:37307)
 is it possible to store or somehow pause training and again resume it with stored weights on the same notebook or on a different machine. 
really Thank you for answering, and thank you for being active and continue being active

**Comment by rojagtap** [comment]
See model checkpointing. I have already used it in the notebook:
```
checkpoint_path = "checkpoints"

ckpt = tf.train.Checkpoint(transformer=transformer, optimizer=optimizer)

ckpt_manager = tf.train.CheckpointManager(ckpt, checkpoint_path, max_to_keep=5)

if ckpt_manager.latest_checkpoint:
    ckpt.restore(ckpt_manager.latest_checkpoint)
    print ('Latest checkpoint restored!!')
```
Just load the last saved checkpoint and fit the model for the next set of epochs.

Ref: https://www.tensorflow.org/api_docs/python/tf/train/Checkpoint

**Comment by SumanthSrungavarapu** [comment]
After 40 epochs loss is down to 2.2, but the output summaries are filled with <stop>. as below

to sell 1st selling house stake in b 39 desh<stop> firm<stop> firm<stop> fund<stop> ceo<stop> ceo<stop> ceo<stop> ceo<stop> ceo<stop> ceo<stop> <stop> fund<stop> fund<stop> <stop> <stop> <stop> <stop> <stop> <stop> fund<stop> fund<stop> fund<stop> <stop> fund<stop> ceo<stop>............

Do you think something is wrong or it just needs more learning?
collab link: https://colab.research.google.com/drive/1JR7bZb5gsmC9zphD-z8oX1gwJVmPk2jm?usp=share_link
